File: src/fpga/layout.py
from __future__ import annotations

from dataclasses import dataclass


@dataclass(frozen=True, slots=True)
class TileShape:
    rows: int
    cols: int
    inner: int

    def validate(self, lhs_len: int, rhs_len: int) -> None:
        expected_lhs = self.rows * self.inner
        expected_rhs = self.inner * self.cols

        if lhs_len != expected_lhs:
            raise ValueError(
                f"lhs tile length mismatch: expected {expected_lhs}, got {lhs_len}"
            )
        if rhs_len != expected_rhs:
            raise ValueError(
                f"rhs tile length mismatch: expected {expected_rhs}, got {rhs_len}"
            )

    @property
    def lhs_size(self) -> int:
        return self.rows * self.inner

    @property
    def rhs_size(self) -> int:
        return self.inner * self.cols

    @property
    def output_size(self) -> int:
        return self.rows * self.cols

File: src/fpga/layout.rs
use anyhow::{Result, bail};

#[derive(Debug, Clone, Copy)]
pub struct TileShape {
    pub rows: usize,
    pub cols: usize,
    pub inner: usize,
}

impl TileShape {
    pub fn validate(self, lhs_len: usize, rhs_len: usize) -> Result<()> {
        if lhs_len != self.rows * self.inner {
            bail!(
                "lhs tile length mismatch: expected {}, got {}",
                self.rows * self.inner,
                lhs_len
            );
        }
        if rhs_len != self.inner * self.cols {
            bail!(
                "rhs tile length mismatch: expected {}, got {}",
                self.inner * self.cols,
                rhs_len
            );
        }
        Ok(())
    }
}
